### DISCUSSION TITLE
টাইপিং করার সময় কিছু সমস্যার সম্মুখিন হয়েছি।

### DISCUSSION THREAD

**Discussion by Mahmudul-hasan613** [OpenBangla/OpenBangla-Keyboard#282] [General]
ওপেন বাংলা কি-বোর্ড কয়েক দিন ব্যবহার করার পরে একটি সমস্যা লক্ষ্য করেছি। সেটি হলো, আমি যদি কোনো শব্দ লেখার পর সেটিকে ডিলিট করার জন্য ctrl + Backspace Shortcut key ব্যবহার করার চেষ্টা করি টাইপ করা অবস্থায়; তখন সেটি কাজ করছে না। এমন কী যদি আমি ctrl + Shift + arrow key ব্যবহার করে কোনো শব্দ নির্বাচন করার চেষ্টা করি টাইপ করার সময়; সেটিও করতে পারি না।

এছাড়া কোনো কিছু লেখার সময় নিচে একটি আন্ডার লাইন দেখা যায়; যেটি টাইপ করার সময় সমস্যা সৃষ্টি করে। সেই আন্ডার লাইনটি মুছার জন্য মানে লেখা শেষ হয়েছে এটি বুঝানোর জন্য Esc, End বাটন বা যেকোনো বিপরীত একটি Arrow key প্রেস করতে হয় যেটি ডিকশনারির দিকে ফোকাস করা থাকে না।

আর একটি বিষয় জানার ছিলো। কোনো কিছু লেখার সময় যেই suggestion দেখানো হয়; সেটি 1, 2, 3 ক্রমিক নং সহ দেখানো হয়। আমার জানার বিষয়টি হলো এমন কোনো keyboard shortcuts আছে কি যেটি ব্যবহার করে আমার প্রয়োজনীয় শব্দটি নির্বাচন করতে পারবো। মানে আমি বলতে চাচ্ছি যে, ধরেণ আমি যদি ক্যালকুলেটর লিখতে চাই, তখন ক্যালকু লেখার পরে 5 নং এ ক্যালকুলেটর শব্দি suggestion হিসেবে দেখিয়েছে, আমি কি চাইলে 5 সংখ্যা ব্যবহার করে আমার কাঙ্ক্ষিত শব্দটি নির্বাচন করতে পারবো...? এমন কোনো keyboard shortcut, openbangla keyboard এ কি আছে...?

**Comment by mominul** [comment]
> সেটি হলো, আমি যদি কোনো শব্দ লেখার পর সেটিকে ডিলিট করার জন্য ctrl + Backspace Shortcut key ব্যবহার করার চেষ্টা করি টাইপ করা অবস্থায়; তখন সেটি কাজ করছে না। 

Ctrl + Backspace সমন্বয়ে আপনি কেমন আচরণ চাচ্ছেন? অর্থাৎ পুরো শব্দটি মুছে যাবে? এরকম ফিচার যুক্ত করা হয়নি, তবে করা যেতে পারে।

> এমন কী যদি আমি ctrl + Shift + arrow key ব্যবহার করে কোনো শব্দ নির্বাচন করার চেষ্টা করি টাইপ করার সময়; সেটিও করতে পারি না।

আপনাকে শুধু arrow কী চাপতে হবে, ctrl + Shift এর সমন্বয়ে নয়।

> এছাড়া কোনো কিছু লেখার সময় নিচে একটি আন্ডার লাইন দেখা যায়; যেটি টাইপ করার সময় সমস্যা সৃষ্টি করে। সেই আন্ডার লাইনটি মুছার জন্য মানে লেখা শেষ হয়েছে এটি বুঝানোর জন্য Esc, End বাটন বা যেকোনো বিপরীত একটি Arrow key প্রেস করতে হয় যেটি ডিকশনারির দিকে ফোকাস করা থাকে না।

এটি আসলে আমাদের হাতে নেই, এটি অ্যাপ্লিকেশান ও ibus এর interaction। আমি এরকম libreoffice এ হতে দেখেছি। 

> আর একটি বিষয় জানার ছিলো। কোনো কিছু লেখার সময় যেই suggestion দেখানো হয়; সেটি 1, 2, 3 ক্রমিক নং সহ দেখানো হয়। আমার জানার বিষয়টি হলো এমন কোনো keyboard shortcuts আছে কি যেটি ব্যবহার করে আমার প্রয়োজনীয় শব্দটি নির্বাচন করতে পারবো। মানে আমি বলতে চাচ্ছি যে, ধরেণ আমি যদি ক্যালকুলেটর লিখতে চাই, তখন ক্যালকু লেখার পরে 5 নং এ ক্যালকুলেটর শব্দি suggestion হিসেবে দেখিয়েছে, আমি কি চাইলে 5 সংখ্যা ব্যবহার করে আমার কাঙ্ক্ষিত শব্দটি নির্বাচন করতে পারবো...? এমন কোনো keyboard shortcut, openbangla keyboard এ কি আছে...?

এমন ফিচার বর্তমানে নেই, তবে implement করা যেতে পারে। ফিচারটি ভালই হবে বলে মনে হচ্ছে 😁 

ধন্যবাদ! 😊

**Reply by rashidi77** [reply]
> Ctrl + Backspace সমন্বয়ে আপনি কেমন আচরণ চাচ্ছেন? অর্থাৎ পুরো শব্দটি মুছে যাবে? এরকম ফিচার যুক্ত করা হয়নি, তবে করা যেতে পারে।

এটি আইবাসের ডিফল্ট আচরণ। আমি আরবি ও ইংরেজিতে এভাবে পুরো শব্দ মুছতে পারি। কিন্তু ওপেনবাংলায় এটা হয় না।

> আপনাকে শুধু arrow কী চাপতে হবে, ctrl + Shift এর সমন্বয়ে নয়।

শুধু এরো কী চেপে সাজেশন নির্বাচনের এই আচরণটি আমার জন্য বিরক্তির কারণ হয়। শুধু arrow চেপে আমরা শব্দের মাঝে কার্সর পরিবর্তন করে অভ্যস্ত। কোনও শব্দের শুরুতে বা মাঝে স্পেলিং কারেকশন করার জন্য এখন আমাদের ডান পাশ থেকে ক্যারেক্টার মোছার প্রয়োজন পড়ছে। যেটিতে ফিক্সড লেআউট ইউজাররা অভ্যস্ত নয়। আমরা সাধারণত arrow কী ব্যবহার করে কার্সর পরিবর্তন করে শব্দ এডিট করে থাকি। এ জন্য সাজেশন নির্বাচনে arrow কী এর ব্যবহারটা বাদ দেয়া উচিত বলে আমি মনে করি। সাজেশন নির্বাচনের জন্য TAB ও Shift TAB তো আছেই।

**Reply by Mahmudul-hasan613** [reply]
rasidi77 ভাইয়ের সাথে আমি সম্পূর্ণ একমত।

**Reply by Mahmudul-hasan613** [reply]
> > এছাড়া কোনো কিছু লেখার সময় নিচে একটি আন্ডার লাইন দেখা যায়; যেটি টাইপ করার সময় সমস্যা সৃষ্টি করে। সেই আন্ডার লাইনটি মুছার জন্য মানে লেখা শেষ হয়েছে এটি বুঝানোর জন্য Esc, End বাটন বা যেকোনো বিপরীত একটি Arrow key প্রেস করতে হয় যেটি ডিকশনারির দিকে ফোকাস করা থাকে না।
> 
> এটি আসলে আমাদের হাতে নেই, এটি অ্যাপ্লিকেশান ও ibus এর interaction। আমি এরকম libreoffice এ হতে দেখেছি।
> 
ibus এর এই interaction টা কী বন্ধ করার কোনো উপায় আছে...? আর এই interation বা ফিচারটির নাম কী...?

এছাড়া আমি নত‌ুন একটি সমস্যা খ‌ুঁজে পেয়েছি, সেটি হলো যখন আমি উব‌ুন্ট‌ুর ফাইল ম্যানেজার চাল‌ু করে কোনো ফাইল খোঁজার চেষ্টা করি, তখন প্রথম অক্ষরটি ইংরেজিতে আসে.... এরপর বাকি অক্ষরগ‌ুলা বাংলা আসে। এতে করে আমার কাঙ্ক্ষিত ফাইল খ‌ুঁজে পেতে একট‌ু কষ্ট হয়...

**Reply by rashidi77** [reply]
> এছাড়া আমি নত‌ুন একটি সমস্যা খ‌ুঁজে পেয়েছি, সেটি হলো যখন আমি উব‌ুন্ট‌ুর ফাইল ম্যানেজার চাল‌ু করে কোনো ফাইল খোঁজার চেষ্টা করি, তখন প্রথম অক্ষরটি ইংরেজিতে আসে.... এরপর বাকি অক্ষরগ‌ুলা বাংলা আসে। এতে করে আমার কাঙ্ক্ষিত ফাইল খ‌ুঁজে পেতে একট‌ু কষ্ট হয়...

এই বাগটা সম্ভবত ফিক্স করা হয়েছ। দেখুন : [f68e918](https://github.com/OpenBangla/OpenBangla-Keyboard/commit/f68e918e2d0928f7a724f9a8395d307a70ffb2c2)

**Reply by Mahmudul-hasan613** [reply]
> > এছাড়া আমি নত‌ুন একটি সমস্যা খ‌ুঁজে পেয়েছি, সেটি হলো যখন আমি উব‌ুন্ট‌ুর ফাইল ম্যানেজার চাল‌ু করে কোনো ফাইল খোঁজার চেষ্টা করি, তখন প্রথম অক্ষরটি ইংরেজিতে আসে.... এরপর বাকি অক্ষরগ‌ুলা বাংলা আসে। এতে করে আমার কাঙ্ক্ষিত ফাইল খুঁজে পেতে ঝামেলা হয়।
> 
> এই বাগটা সম্ভবত ফিক্স করা হয়েছ। দেখুন : [f68e918](https://github.com/OpenBangla/OpenBangla-Keyboard/commit/f68e918e2d0928f7a724f9a8395d307a70ffb2c2)

কিন্তু এই সমস্যা টি এখনো আমার হচ্ছে। আমি জাতীয় কি-বোর্ড লে-আউটে লিখি, ধরেন আমি ডাউনলোড ফোল্ডার খুঁজতে চাই; তো ড থাকে e বাটনে, যখন আমি ড লিখতে চাই তখন সেটা e হিসাবে আসে, এরপর যেই অক্ষর গুলো লিখি সেগুলো বাংলা অক্ষর হিসেবে আসে। ফাইল ম্যানেজারে কোনো ফোল্ডার খুঁজতে গেলে তখন আমি উক্ত সমস্যার সম্মুখীন হই। কিন্তু অন্য কোথায় লেখার সময় এই সমস্যা টা হয় না। মানে libre office এ টাইপ করার সময় এই সমস্যা হয় না। শুধু ফাইল ম্যানেজারের ক্ষেত্রেই এই সমস্যা টা হয়....

**Reply by rashidi77** [reply]
> কিন্তু এই সমস্যা টি এখনো আমার হচ্ছে।

স্বাভাবিক। পরবর্তি আপডেট আসা পর্যন্ত আপনার অপেক্ষা করতে হবে।  বা চাইলে আপনি cমেনুয়ালি প্যাচগুলা এপ্লাই করে dনিতে পারেন। 

[আমার মনে হয় ওপেনবাংলা কীবোর্ডের বাগ-ফিক্সগুলা ছোটখাট আপডেটের মাধ্যমে পুশ করা উচিত, কিন্তু এর জন্য দরকার একটা একটিভ রিপো]

**Reply by ahmubashshir** [reply]
@mominul bro, you can roll-out minor versions each ~5-20 commits, see https://semver.org/

**Comment by mominul** [comment]
>>  Ctrl + Backspace সমন্বয়ে আপনি কেমন আচরণ চাচ্ছেন? অর্থাৎ পুরো শব্দটি মুছে যাবে? এরকম ফিচার যুক্ত করা হয়নি, তবে করা যেতে পারে।

> এটি আইবাসের ডিফল্ট আচরণ। আমি আরবি ও ইংরেজিতে এভাবে পুরো শব্দ মুছতে পারি। কিন্তু ওপেনবাংলায় এটা হয় না।

@rashidi77  এই `Ctrl + Backspace` এর ফিচারটি implement করার ইচ্ছা আছে আমার, এটা নিয়ে কাজ শুরু করব ইনশাল্লাহ।

> এছাড়া আমি নত‌ুন একটি সমস্যা খ‌ুঁজে পেয়েছি, সেটি হলো যখন আমি উব‌ুন্ট‌ুর ফাইল ম্যানেজার চাল‌ু করে কোনো ফাইল খোঁজার চেষ্টা করি, তখন প্রথম অক্ষরটি ইংরেজিতে আসে.... এরপর বাকি অক্ষরগ‌ুলা বাংলা আসে। এতে করে আমার কাঙ্ক্ষিত ফাইল খ‌ুঁজে পেতে একট‌ু কষ্ট হয়...

@Mahmudul-hasan613 এই বাগটি সম্পর্কে আমি আগে অবগত ছিলাম না, এবং এটি আগে রিপোর্ট করা হয়নি। তাই @rashidi77 ভাইয়ের লিঙ্কে দেয়া ফিক্স অন্য একটি ফিচারের বাগফিক্স(যেটা নতুন ভার্সনে আসবে)। তবে আপনার বর্ণনা থেকে যা বুঝতে পারছি এটি ibus এবং ফাইল ম্যনেজারের মধ্যে interaction এ সমস্যা, সেক্ষেত্রে আমাদের কিছু করার থাকবে না, তবুও আমি reproduce করে দেখব যে আমাদের কোন কিছু করার আছে কিনা। 
তবে নতুন ভার্সনে fcitx এর সাপোর্ট আসতে চলেছে, যেটা ব্যবহার করে দেখলাম এই বাগটি হচ্ছে না ফাইল ম্যানেজারের ক্ষেত্রে। তাই ibus এর জন্য যারা সমস্যার সম্মুখীন হচ্ছেন তারা ভবিষ্যতে fcitx ইউজ করলে এসব ভোগান্তি থেকে মুক্তি পাবেন আশা করি।

>শুধু এরো কী চেপে সাজেশন নির্বাচনের এই আচরণটি আমার জন্য বিরক্তির কারণ হয়। শুধু arrow চেপে আমরা শব্দের মাঝে কার্সর পরিবর্তন করে অভ্যস্ত। কোনও শব্দের শুরুতে বা মাঝে স্পেলিং কারেকশন করার জন্য এখন আমাদের ডান পাশ থেকে ক্যারেক্টার মোছার প্রয়োজন পড়ছে। যেটিতে ফিক্সড লেআউট ইউজাররা অভ্যস্ত নয়। আমরা সাধারণত arrow কী ব্যবহার করে কার্সর পরিবর্তন করে শব্দ এডিট করে থাকি। এ জন্য সাজেশন নির্বাচনে arrow কী এর ব্যবহারটা বাদ দেয়া উচিত বলে আমি মনে করি। সাজেশন নির্বাচনের জন্য TAB ও Shift TAB তো আছেই।

@rashidi77 যেহেতু আমাদের সামনে একটি লিস্ট দেখানো হয় এবং আমরা লিস্ট এরো কি ব্যবহার করে সিলেক্ট করতে অভ্যস্ত তাই অনেকের কাছেই কিন্তু এরো দিয়ে সাজেশন পরিবর্তন করা ব্যাপারটা লজিকাল লাগবে এবং লাগার কথা।  অনেকে এক্ষেত্রে ট্যাবের ব্যাবহার নাও জানতে পারেন। 
আপনার যদি লেফট রাইট এরো ব্যবহার করতে হয় তাইলে আপনি Vertical suggestion লিস্ট সিলেক্ট করতে পারবেন সেটিং থেকে, তখন সাজেশন সিলেকশনে আপ এবং ডাউন কি ব্যবহৃত হবে আর লেফট রাইট এরো কার্সর সরাতে ইউজ হবে।

>[আমার মনে হয় ওপেনবাংলা কীবোর্ডের বাগ-ফিক্সগুলা ছোটখাট আপডেটের মাধ্যমে পুশ করা উচিত, কিন্তু এর জন্য দরকার একটা একটিভ রিপো]

> @mominul bro, you can roll-out minor versions each ~5-20 commits, see https://semver.org/

@rashidi77 @ahmubashshir হ্যাঁ এরকম করা যেত যদিনা এই ভার্সনে নতুন একটা backend, fcitx যুক্ত হত। এটি একটি বড় পরিবর্তন যার জন্য আমাদের পুরো packaging ও installation strategy বদলে গেছে। আর fcitx backend এ পুরোপুরি feature parity এনেই রিলিজ দেয়ার ইচ্ছা আমার, আমি এখন সবসময় fcitx ব্যবহার করা শুরু করেছি যেন তাড়াতাড়ি সেটি করতে পারি। 
আর যদি কোন সেরিয়াস কিংবা critical বাগ পাওয়া যায় তখন তো নতুন ভার্শন রিলিজ করতে হবেই, সেজন্য এবার সব নতুন ডেভেলপমেন্ট একটা আলাদা `develop` ব্রাঞ্চে হচ্ছে। যেন পূর্ববর্তী কোডর উপর সহজেই patch অ্যাপ্লাই করা যায়।

ধন্যবাদ
